File: testcases/paper/Generics/Generics.cpp
template<typename Number>
class Complex {
private:
    Number re;
    Number im;

public:
    static const Complex<Number> I;

    Complex(Number re = Number{}, Number im = Number{}) : re(re), im(im) {}

    void clear_imag() {
        im = Number{};
    }

    void clear_real() {
        re = Number{};
    }

    void set(const Complex<Number>& val) {
        re = val.re;
        im = val.im;
    }

    Complex operator+(const Complex& rhs) const {
        return Complex(re + rhs.re, im + rhs.im);
    }
};

template<typename Number>
const Complex<Number> Complex<Number>::I(Number{}, Number{1});

File: testcases/paper/Generics/Generics.java
public class Generics<Number> {
    private Number re;
    private Number im;
    private static final Generics<?> I = new Generics<>(null, null); // Placeholder, requires actual zero and one representations

    public Generics(Number re, Number im) {
        this.re = re;
        this.im = im;
    }

    public void clearImag(Addable<Number> zeroRep) {
        this.im = zeroRep.zero();
    }

    public void clearReal(Addable<Number> zeroRep) {
        this.re = zeroRep.zero();
    }

    public void set(Generics<Number> val) {
        this.re = val.re;
        this.im = val.im;
    }

    public Generics<Number> add(Generics<Number> rhs, Addable<Number> adder) {
        return new Generics<>(adder.add(this.re, rhs.re), adder.add(this.im, rhs.im));
    }

    public interface Addable<N> {
        N add(N left, N right);
        N zero();
    }
}
